agent/reporter.go

```diff
@@ -102,27 +124,30 @@ func (sr *statusReporter) run(ctx context.Context) {
 			return
 		}
 
-		for taskID, status := range sr.statuses {
-			delete(sr.statuses, taskID) // delete the entry, while trying to send.
-
-			sr.mu.Unlock()
-			err := sr.reporter.UpdateTaskStatus(ctx, taskID, status)
-			sr.mu.Lock()
-
-			// reporter might be closed during UpdateTaskStatus call
-			if sr.closed {
-				return
-			}
-
-			if err != nil {
-				log.G(ctx).WithError(err).Error("status reporter failed to report status to agent")
-
-				// place it back in the map, if not there, allowing us to pick
-				// the value if a new one came in when we were sending the last
-				// update.
-				if _, ok := sr.statuses[taskID]; !ok {
-					sr.statuses[taskID] = status
-				}
+		// swap out the statuses map, so we can release the lock while we
+		// do the batch
+		statuses := sr.statuses
+		sr.statuses = map[string]*api.TaskStatus{}
```

Comment:
Because this copies the old status, an out of date copy has the potential to come through.

Reply:
sure, but that was the case before this change as well. if a new status comes in, then it will get addressed in the next go-round.